code/modules/admin/tag.dm

```diff
@@ -0,0 +1,128 @@
+/**
+ * Creates a weakref of the target datum and inserts it into [/datum/admins/var/tagged_datums], for later reference.
+ *
+ * Note that if the original target is deleted and another datum is put in the same memory address, tthis
+ *
+ * Arguments:
+ * * target_datum - The datum you want to create a tag for. Should be the datum itself, not a weakref
+ */
+/datum/admins/proc/add_tagged_datum(datum/target_datum)
+	var/datum/weakref/new_weakref = WEAKREF(target_datum)
+
+	if(LAZYFIND(tagged_datums, new_weakref))
+		new_weakref = null // needed?
+		to_chat(owner, span_warning("[target_datum] is already tagged!"))
+		return
+
+	LAZYADD(tagged_datums, new_weakref)
+	RegisterSignal(target_datum, COMSIG_PARENT_QDELETING, .proc/handle_tagged_del)
+	to_chat(owner, span_notice("[target_datum] has been tagged."))
+
+/// Get ahead of the curve with deleting
+/datum/admins/proc/handle_tagged_del(datum/source)
+	SIGNAL_HANDLER
+
+	if(owner)
+		to_chat(owner, span_boldnotice("Tagged datum [source] ([source.type]) has been deleted."))
+	remove_tagged_datum(source, silent = TRUE)
+
+/**
+ * Attempts to remove the specified datum from [/datum/admins/var/tagged_datums] if it exists
+ *
+ * The argument can be either the datum itself or the specific weakref used to link to it from this admins datum.
+ *
+ * Arguments:
+ * * target_datum - The datum you want to remove from the tagged_datums list. Can also be the specific weakref used in said list.
+ * * silent - If TRUE, won't print messages to the owner's chat
+ */
+/datum/admins/proc/remove_tagged_datum(datum/target_datum, silent=FALSE)
+	if(isnull(target_datum))
+		return
+
+	if(!istype(target_datum)) // for handling memory REF(X)'s to the target datum
+		target_datum = locate(target_datum)
+
+	if(isweakref(target_datum)) // no need to iterate and resolve each weakref
+		if(LAZYFIND(tagged_datums, target_datum))
+			LAZYREMOVE(tagged_datums, target_datum)
+			if(!silent)
+				to_chat(owner, span_notice("[target_datum] has been untagged."))
+		else if(!silent)
+			to_chat(owner, span_warning("[target_datum] was not already tagged."))
+		return
+	else // we have to actually resolve each tagged weakref to see which one is ours
+		for(var/datum/weakref/iter_weakref as anything in tagged_datums)
+			if(iter_weakref.resolve() == target_datum)
+				LAZYREMOVE(tagged_datums, iter_weakref)
+				UnregisterSignal(target_datum, COMSIG_PARENT_QDELETING)
+				if(!silent)
+					to_chat(owner, span_notice("[target_datum] has been untagged."))
+				return
```

Comment:
You don't actually need this loop. Each datum will only ever have one weakref attached to it, so something like this would be valid.

I have a feeling that'd clean up the bit above, but it may become a bit meaningless if you decide to ditch weakrefs 
```suggestion
	else // we have to actually resolve each tagged weakref to see which one is ours
		var/datum/weakref/ref_to_us = WEAKREF(target_datum)
		if(!(ref_to_us in tagged_datums)
			//Failure case
			return  
		LAZYREMOVE(tagged_datums, ref_to_us)
		UnregisterSignal(target_datum, COMSIG_PARENT_QDELETING)
		if(!silent)
			to_chat(owner, span_notice("[target_datum] has been untagged."))
		return
```